lib/semantics/expression.cc

```diff
@@ -1798,51 +1826,44 @@ MaybeExpr ExpressionAnalyzer::Analyze(const parser::Expr::Parentheses &x) {
   return std::nullopt;
 }
 
-MaybeExpr ExpressionAnalyzer::Analyze(const parser::Expr::UnaryPlus &x) {
-  MaybeExpr value{Analyze(x.v.value())};
-  if (value.has_value()) {
-    if (!std::visit(
-            [&](const auto &y) {
-              using yTy = std::decay_t<decltype(y)>;
-              if constexpr (std::is_same_v<yTy, BOZLiteralConstant>) {
-                // allow and ignore +Z'1', it's harmless
-                return true;
-              } else if constexpr (!IsNumericCategoryExpr<yTy>()) {
-                Say("Operand of unary + must have numeric type"_err_en_US);
-                return false;
-              } else {
-                return true;
-              }
-            },
-            value->u)) {
-      return std::nullopt;
+static MaybeExpr NumericUnaryHelper(ExpressionAnalyzer &context,
+    NumericOperator opr, const parser::Expr::IntrinsicUnary &x) {
+  ArgumentAnalyzer analyzer{context};
+  analyzer.Analyze(x.v);
+  if (analyzer.fatalErrors()) {
+    return std::nullopt;
+  } else if (analyzer.IsIntrinsicNumeric()) {
+    if (opr == NumericOperator::Add) {
+      return analyzer.GetAsExpr(0);
+    } else {
+      return Negation(context.GetContextualMessages(), analyzer.GetAsExpr(0));
     }
+  } else {
+    return analyzer.TryDefinedOp(AsFortran(opr),
+        "Operand of unary %s must be numeric; have %s"_err_en_US);
   }
-  return value;
+}
+
+MaybeExpr ExpressionAnalyzer::Analyze(const parser::Expr::UnaryPlus &x) {
+  return NumericUnaryHelper(*this, NumericOperator::Add, x);
 }
 
 MaybeExpr ExpressionAnalyzer::Analyze(const parser::Expr::Negate &x) {
-  if (MaybeExpr operand{Analyze(x.v.value())}) {
-    return Negation(GetContextualMessages(), std::move(*operand));
-  }
-  return std::nullopt;
+  return NumericUnaryHelper(*this, NumericOperator::Subtract, x);
 }
 
 MaybeExpr ExpressionAnalyzer::Analyze(const parser::Expr::NOT &x) {
-  if (MaybeExpr operand{Analyze(x.v.value())}) {
-    return std::visit(
-        common::visitors{
-            [](Expr<SomeLogical> &&lx) -> MaybeExpr {
-              return {AsGenericExpr(LogicalNegation(std::move(lx)))};
-            },
-            [&](auto &&) -> MaybeExpr {
-              Say("Operand of .NOT. must be LOGICAL"_err_en_US);
-              return std::nullopt;
-            },
-        },
-        std::move(operand->u));
+  ArgumentAnalyzer analyzer{*this};
+  analyzer.Analyze(x.v);
+  if (analyzer.fatalErrors()) {
+    return std::nullopt;
+  } else if (analyzer.IsIntrinsicLogical()) {
+    return AsGenericExpr(
+        LogicalNegation(std::get<Expr<SomeLogical>>(analyzer.GetAsExpr(0).u)));
+  } else {
+    return analyzer.TryDefinedOp(
+        ".not.", "Operand of %s must be LOGICAL; have %s"_err_en_US);
```

Comment:
Why is `.not.` in lower case while `LOGICAL` is upper case?  (rhetorical question, the usage just seems inconsistent)

Reply:
The generic symbol is created with the name "operator(.not.)" in lower case, from the cooked source. But when it's referenced in an error message it is converted to ".NOT.".